src/solvers/dg/2d/dg.jl

```diff
@@ -2085,9 +2152,66 @@ function calc_mortar_flux!(surface_flux_values, dg::Dg2D, mortar_type::Val{:l2},
     calc_fstar!(fstar_upper, u_upper, m, orientations, dg)
     calc_fstar!(fstar_lower, u_lower, m, orientations, dg)
 
-    mortar_fluxes_to_elements!(surface_flux_values, dg, mortar_type, m,
-                               fstar_upper, fstar_lower)
-  end
+    if dg.use_flux_correction
+      mortar_fluxes_to_elements_lgl!(surface_flux_values, dg, mortar_type, m,
+                                     fstar_upper, fstar_lower)
+    else
+      mortar_fluxes_to_elements!(surface_flux_values, dg, mortar_type, m,
+                                fstar_upper, fstar_lower)
+    end
+  end
+
+  # Chan et al. flux correction
+  @timeit timer() "flux correction" if dg.use_flux_correction
+    # Notation:
+    # - u_large[v,j]       = Uⱼᶠ solution values on large face
+    # - u_large_upper[v,p] = Uₚᵐ solution values on *upper* mortar
+    # - u_large_lower[v,p] = Uₚᵐ solution values on *lower* mortar
+    # - fstar_upper_correction[v,j,p] = Fᵥ(Uⱼᶠ, Uₚᵐ) for *upper* mortar
+    # - fstar_lower_correction[v,j,p] = Fᵥ(Uⱼᶠ, Uₚᵐ) for *lower* mortar
+
+    # Threads.@threads for m in 1:dg.n_l2mortars
+    for m in 1:dg.n_l2mortars
+      # Choose thread-specific pre-allocated container
+      fstar_upper = fstar_upper_threaded[Threads.threadid()]
+      fstar_lower = fstar_lower_threaded[Threads.threadid()]
+      fstar_upper_correction = fstar_upper_correction_threaded[Threads.threadid()]
+      fstar_lower_correction = fstar_lower_correction_threaded[Threads.threadid()]
+
+      large_element_id = dg.l2mortars.neighbor_ids[3, m]
+
+      if dg.l2mortars.large_sides[m] == 1 # -> large element on left side
+        if dg.l2mortars.orientations[m] == 1
+          # L2 mortars in x-direction
+          direction = 2
+          u_large = view(dg.elements.u, :, nnodes(dg), :, large_element_id)
+        else
+          # L2 mortars in y-direction
+          direction = 4
+          u_large = view(dg.elements.u, :, :, nnodes(dg), large_element_id)
+        end
+        u_large_upper = view(u_upper, 1, :, :, m)
+        u_large_lower = view(u_lower, 1, :, :, m)
```

Comment:
That's not type-stable; the type of `u_large` cannot be determined at compile time and you will get dynamic dispatch. That's why I wrote 
https://github.com/trixi-framework/Trixi.jl/blob/master/src/solvers/dg/2d/dg.jl#L1769-L1777
as it is. The following changes should help (disclaimer: I haven't tested them, I'm just guessing).
```suggestion
        u_large_upper = view(u_upper, 1, :, :, m)
        u_large_lower = view(u_lower, 1, :, :, m)
        if dg.l2mortars.orientations[m] == 1
          # L2 mortars in x-direction
          direction = 2
          u_large = view(dg.elements.u, :, nnodes(dg), :, large_element_id)
          calc_flux_correction!(surface_flux_values, dg, u_large, u_large_upper, u_large_lower,
                                m, direction, large_element_id,
                                fstar_upper_correction, fstar_lower_correction)
        else
          # L2 mortars in y-direction
          direction = 4
          u_large = view(dg.elements.u, :, :, nnodes(dg), large_element_id)
          calc_flux_correction!(surface_flux_values, dg, u_large, u_large_upper, u_large_lower,
                                m, direction, large_element_id,
                                fstar_upper_correction, fstar_lower_correction)
        end
```